### PR TITLE
Dev oleguer

### LINKED ISSUE #7: Write objective function (loss)

### LINKED ISSUE #10: Integrator
We need to implement the **leapfrog integrator** from the HNN.
We should also implement **euler integration** and **runge-kutta rk4** integrator.

I'm thinking of having an integrator class and pass it to the model. Something like:

```python
integrator = Integrator(method="euler")
q_next, p_next = integrator.step(q_now, p_now, HNN, dt)
```

What do you guys think @charlio23  @fedetask

### PR DESCRIPTION
**Pull request by OleguerCanal**
HGN training pipeline + integrator + loss function

Missing:
- [x] KL divergence in loss
- [x] Saving/Loading of model architecture and params
- [x] Integrating new dataloader (env and models still not connected)

### CODE REVIEW

File: environments/environment_factory.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+"""Environment factory class. Given a valid environment name and its constructor args, returns an instantiation of it
+"""
+from .environments import Environment
+# TODO(Oleguer): Fix this, its not very nice to have to import all classes
+from .pendulum import Pendulum
+# from spring import Spring
```
Comment (charlio23) [inline environments/environment_factory.py]: Should we import all the implemented environments here?

File: environments/pendulum.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ def __init__(self, mass, length, g, q=None, p=None):
         self.mass = mass
         self.length = length
         self.g = g
-        super().__init__(q=q, p=p)
+        super(Pendulum, self).__init__(q=q, p=p)
```
Comment (charlio23) [inline environments/pendulum.py]: Why do you call `super `using `Pendulum` and `self`? Does this needs changes in the `Environment` class?

File: hamiltonian_generative_network.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import torch
+
+from utilities.integrator import Integrator
+from utilities.hgn_result import HgnResult
+
+
+class HGN():
+    """Hamiltonian Generative Network model.
+    
+    This class models the HGN and allows implements its training and evaluation.
```
Comment (charlio23) [inline hamiltonian_generative_network.py]: allows implementing 🍔

File: hamiltonian_generative_network.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            q, p = self.integrator.step(q=q, p=p, hnn=self.hnn)
+            prediction.append_state(q=q, p=p)
+
+            # Compute state reconstruction
+            x_reconstructed = self.decoder(q=q)
+            prediction.append_reconstruction(x_reconstructed)
+        return prediction
+
+    def fit(self, rollouts):
+        """Perform a training step with the given rollouts batch.
```
Comment (charlio23) [inline hamiltonian_generative_network.py:99]: at some point in the training step before the forward pass we should have the models uploaded to the GPU as well as the batch that has to be forwarded.

**Review by charlio23 — COMMENTED** [review]
Nice implementation! Just one important note. In the training script we need to upload the netoworks and the batches to the GPU.

A good practice would be to have a boolean in the yaml file to select if you want to use GPU or CPU computations.

File: environments/environment_factory.py (hunk 1 above)
Comment (OleguerCanal) [reply environments/environment_factory.py]: yes, not ideal

File: environments/pendulum.py (hunk 2 above)
Comment (OleguerCanal) [reply environments/pendulum.py]: Sorry, I think its not necessary. I'll remove it

File: hamiltonian_generative_network.py (hunk 4 above)
Comment (OleguerCanal) [reply hamiltonian_generative_network.py:99]: Good point, I'll will ad a check that uses gpu if available

**Comment by OleguerCanal** [comment]
Batch size bigger than one doesnt work (integrator fails), for the rest it seems that the pipeline works without crashing

File: environments/environment_factory.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+"""Environment factory class. Given a valid environment name and its constructor args, returns an instantiation of it
+"""
+from .environments import Environment
+# TODO(Oleguer): Fix this, its not very nice to have to import all classes
+from .pendulum import Pendulum
+from .spring import Spring
+
```
Comment (fedetask) [inline environments/environment_factory.py]: Can we use `sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))`? The relative import is not working for me. 
```suggestion
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from environments import Environment
# TODO(Oleguer): Fix this, its not very nice to have to import all classes
from pendulum import Pendulum
from spring import Spring

```

File: environments/pendulum.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import numpy as np
 
-from environments import Environment, visualize_rollout
+from .environments import Environment, visualize_rollout
```
Comment (fedetask) [inline environments/pendulum.py]: Same as above

File: environments/spring.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import numpy as np
 
-from environments import Environment, visualize_rollout
+from .environments import Environment, visualize_rollout
```
Comment (fedetask) [inline environments/spring.py]: Same as above

File: hamiltonian_generative_network.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        """Get the prediction of the HGN for a given rollout_batch of n_steps.
+
+        Args:
+            rollout_batch (torch.Tensor(N, C, H, W)): Image sequence of the system evolution concatenated along the channels' axis.
+            n_steps (integer, optional): Number of guessed steps, if None will match seq_len. Defaults to None.
+
+        Returns:
+            HgnResult: Bundle of the intermediate and final results of the HGN output.
+        """
+        rollout_batch = concat_rgb(rollout_batch)
```
Comment (fedetask) [inline hamiltonian_generative_network.py:68]: This should be removed once the environments output the correct shape, right?

File: train.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+from environments.environment_factory import EnvFactory
+from hamiltonian_generative_network import HGN
+from networks.inference_net import EncoderNet, TransformerNet
+from networks.hamiltonian_net import HamiltonianNet
+from networks.decoder_net import DecoderNet
+import utilities
+
+params_file = "experiment_params/default.yaml"
+
+if __name__ == "__main__":
```
Comment (fedetask) [inline train.py]: At some point we will need to have this in a function, e.g. if we want to run multiple experiments automatically

File: train.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    decoder = DecoderNet(
+        in_channels=params["networks"]["transformer"]["out_channels"],
+        out_channels=params["rollout"]["n_channels"],
+        **params["networks"]["decoder"])
+
+    # Define HGN integrator
+    integrator = utilities.integrator.Integrator(
+        delta_t=params["rollout"]["delta_time"],
+        method=params["integrator"]["method"])
+
```
Comment (fedetask) [inline train.py:54]: Maybe we'll need to have this in a function of HGN, something like a static function `HGN.from_params()` that returns an `HGN` instance with all the attributes instantiated from the yaml configuration

File: utilities/integrator.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        Returns:
+            tuple(torch.Tensor, torch.Tensor): Position and momentum time derivatives: dq_dt, dp_dt.
+        """
+        # Compute energy of the system
+        energy = hnn(q=q, p=p)
+
+        # dq_dt = dH/dp
+        dq_dt = torch.autograd.grad(energy, p, create_graph=True, retain_graph=True)[0]
+        # dp_dt = -dH/dq
+        dp_dt = -torch.autograd.grad(energy, q, create_graph=True, retain_graph=True)[0]
```
Comment (fedetask) [inline utilities/integrator.py]: Change this as we discussed for allowing minibatches
```suggestion
        dq_dt = torch.autograd.grad(energy, p, create_graph=True, retain_graph=True,
            grad_outputs=torch.ones_like(energy))[0]
        # dp_dt = -dH/dq
        dp_dt = -torch.autograd.grad(energy, q, create_graph=True, retain_graph=True,
            grad_outputs=torch.ones_like(energy))[0]
```

**Review by fedetask — CHANGES_REQUESTED** [review]
Nice! There are a few small things to be fixed, then we'll need to move the training into a more structured function/class.

File: environments/environment_factory.py (hunk 5 above)
Comment (OleguerCanal) [reply environments/environment_factory.py]: os.path.dirname(os.path.dirname( 2 times is an error or intended?

File: hamiltonian_generative_network.py (hunk 8 above)
Comment (OleguerCanal) [reply hamiltonian_generative_network.py:68]: Yes

File: environments/environment_factory.py (hunk 5 above)
Comment (fedetask) [reply environments/environment_factory.py]: It's intended because dirname takes the name of the directory of the file. So if `__file__ = 'Hamiltonian-Generative-Networks/networks/inference_net.py'`, then `os.path.dirname(__file__) = Hamiltonian-Generative-Networks/networks/`, and using it twice gives `os.path.dirname(os.path.dirname(__file__)) = 'Hamiltonian-Generative-Networks'` which is the root of the project.

**Review by fedetask — APPROVED** [review]
Everything is good now
